src/shared/common/vector_functions.h

```diff
@@ -131,6 +128,15 @@ inline RealType VecdSquareNorm(const Eigen::Matrix<RealType,Dimension,1>& vec) {
     return vec.squaredNorm();
 }
 
+template<class VecType, std::size_t ...Index>
+inline auto VecdFoldingProd_impl(const VecType& vec, std::index_sequence<Index...>) {
+    return ( vec[Index] * ... );
+}
+template<class Type, int Dimension>
+inline Type VecdFoldProduct(const sycl::vec<Type,Dimension>& vec) {
+    return VecdFoldingProd_impl(vec, std::make_index_sequence<Dimension>());
+}
+
 /* SYCL memory transfer utilities */
```

Comment:
Interesting, you introduce function_impl.

Reply:
What is this for?

Reply:
> - Allocated size of CellLinkedListKernel::index_head_list_ is now generalized for any dimension, instead of defaulting to the 2d case

I used it for this point, essentially `index_head_list_` needs to allocate `all_cells[0] * all_cells[1]` elements for 2d builds and `all_cells[0] * all_cells[1] * all_cells[2]` in the 3d case. I used a [fold expression](https://en.cppreference.com/w/cpp/language/fold) to generalize the product of all elements based on the vector dimension. Having a second function impl is required for expression folding in order to create the index sequence. There are two alternatives to this approach:
1. Using `std::accumulate`, however `sycl::vec` does not provide iterators (`begin` and `end`)
2. Using a simple ternary operator `all_cells[0] * all_cells[1] * (Dimension == 3 : all_cells[2] : 1)`. Which is what `VecdFoldProduct` is computing implicitly, so that there is no need to remember to add the whole expression every time the allocation size of `index_head_list_` is needed